### PR TITLE
Introduces ble_writer_switch

### PR DESCRIPTION
**Pull request by rbaron**
Hi there,

This is a component I implemented for a personal use case: toggling my custom solar-powered balcony lights. The lights expose a BLE service that clients can write to in order to turn them on/off. There seems to be some interest in this functionality (e.g.: esphome/feature-requests/issues/1460 and maybe can be extended to tackle esphome/feature-requests/issues/454), so I'm submitting this draft PR hoping it might be useful for more users.

This is a basic first stab at this new switch platform. An immediate improvement is to implement support for a `lambda` so that the actual bytes written to the characteristic are configurable. For now they are simply `0x00` (switch turned off) and `0x01` (switch turned on).
 
# What does this implement/fix? 
This is a switch that triggers a write to a specified BLE characteristic
when toggled.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1937

## Test Environment
- [ ] ESP32
- [X] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
esphome:
  name: solarleds-bridge
  platform: ESP32
  board: lolin32

esp32_ble_tracker:

ble_client:
  - mac_address: DD:B6:AF:13:23:A8
    id: solarleds_ble_client

switch:
  - platform: ble_writer_switch
    ble_client_id: solarleds_ble_client
    name: "Solarleds Switch"
    service_uuid: F61E3BE9-2826-A81B-970A-4D4DECFABBAE
    characteristic_uuid: 6490FAFE-0734-732C-8705-91B653A081FC
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ble_writer_switch/switch.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+CONFIG_SCHEMA = (
+    switch.SWITCH_SCHEMA.extend(
+        {
+            cv.GenerateID(): cv.declare_id(BLEWriterSwitch),
+            cv.Required(CONF_SERVICE_UUID): esp32_ble_tracker.bt_uuid,
+            cv.Required(CONF_CHARACTERISTIC_UUID): esp32_ble_tracker.bt_uuid,
+            cv.Optional(CONF_INVERTED): cv.invalid(
+                "BLE writer switche does not support inverted mode"
```
Comment (oli101010) [inline esphome/components/ble_writer_switch/switch.py]: Typo:

```suggestion
                "BLE writer switch does not support inverted mode"
```

**Comment by ghost** [comment]
Hi, i have just installed latest esphome dev but i obtain:
 Platform not found: 'switch.ble_writer_switch'.
There is a way for testing this new library?

**Comment by randybb** [comment]
@niahane 
```yaml
external_components:
  - source: github://pr#2776
    components: ["ble_writer_switch"]
```

**Comment by ghost** [comment]
Thank you, now seems to load this component.
I am trying to move my switchbot but i dont know where put the hexadecimal write command 0x570100 in characteristic uuid:
Can you help me please?

esp32_ble_tracker:
ble_client:
  - mac_address: EC:5F:3A:CA:70:1D
    id: switchbot_a

switch:
  - platform: ble_writer_switch
    ble_client_id: switchbot_a
    name: "Switchbot test"
    service_uuid: cba20d00-224d-11e6-9fb8-0002a5d5c51b
    characteristic_uuid: cba20003-224d-11e6-9fb8-0002a5d5c51b
    descriptor_uuid: ???

**Review by OttoWinter — COMMENTED** [review]
LGTM, can you fix the typo?

File: esphome/components/ble_writer_switch/switch.py (hunk 1 above)
Comment (rbaron) [reply esphome/components/ble_writer_switch/switch.py]: Fixed.

**Comment by rbaron** [comment]
- Fixed typo
- Rebase against `dev` branch
- Moved the definition of `CONF_CHARACTERISTIC_UUID` to `const.py` and updated similar files to satisfy new lint rules

If this gets merged, we need to re-open the esphome-docs PR in [esphome/esphome-docs#1677](https://github.com/esphome/esphome-docs/pull/1677). It seems like I cannot do it myself.

**Comment by jesserockz** [comment]
@rbaron you will need to create a new docs PR as GitHub will not allow me to reopen the docs PR either due to branch changes.

**Comment by rbaron** [comment]
Hi @jesserockz no worries, I created a new one in esphome/esphome-docs#1937 and updated the original pull request message to reference it instead.

**Comment by WeekendWarrior1** [comment]
Not to seem unthankful, because this functionality is incredible, but would this maybe make more sense as an action of `ble_client` ?
Considering that BLE is basically only read/write/notify/indicate, this would lay a framework for maybe the other options (at the least, `READ`)

eg:
```yaml
esphome:
  name: solarleds-bridge
  platform: ESP32
  board: lolin32

esp32_ble_tracker:

ble_client:
  - mac_address: DD:B6:AF:13:23:A8
    id: solarleds_ble_client

switch:
  - platform: template
    name: "Template Switch"
    turn_on_action:
      - ble_client.write:
        id: solarleds_ble_client
        service_uuid: F61E3BE9-2826-A81B-970A-4D4DECFABBAE
        characteristic_uuid: 6490FAFE-0734-732C-8705-91B653A081FC
        value: 0x01
    turn_off_action:
      - ble_client.write:
        id: solarleds_ble_client
        service_uuid: F61E3BE9-2826-A81B-970A-4D4DECFABBAE
        characteristic_uuid: 6490FAFE-0734-732C-8705-91B653A081FC
        value: 0x00
```

**Comment by rbaron** [comment]
Hi @WeekendWarrior1,

I think your suggestion makes a lot of sense. I got curious and took a stab at implementing it in [rbaron/esphome](https://github.com/rbaron/esphome/tree/ble_write_action) (`ble_write_action` branch). I tested it and it seems to work as well as this PR, and I think it's slightly more flexible. Would you be interested in helping me test this out?

There are a few things missing (I left two TODOs there):
- [ ]  Handling 16- and 32-byte-long BLE UUIDs
- [ ] Handling arbitrary `data`. For this proof of concept it's just a single `uint8_t`, but it would be nice to support any byte array

The config I'm using for testing is:

```yaml
esphome:
  name: solarleds-bridge-action
  platform: ESP32
  board: lolin32

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

mqtt:
  broker: !secret mqtt_broker_hostname
  username: mqttuser
  password: !secret mqtt_password

esp32_ble_tracker:

ble_client:
  - mac_address: DD:B6:AF:13:23:A8
    id: solarleds_ble_client

switch:
  - platform: template
    name: "Solarleds Template Switch"
    optimistic: true
    turn_on_action:
      - ble_client.ble_write:
          id: solarleds_ble_client
          service_uuid: F61E3BE9-2826-A81B-970A-4D4DECFABBAE
          characteristic_uuid: 6490FAFE-0734-732C-8705-91B653A081FC
          value: 0x01
    turn_off_action:
      - ble_client.ble_write:
          id: solarleds_ble_client
          service_uuid: F61E3BE9-2826-A81B-970A-4D4DECFABBAE
          characteristic_uuid: 6490FAFE-0734-732C-8705-91B653A081FC
          value: 0x00

logger:
```

With this config I can turn my BLE balcony lights on/off just as well.

I'd also like to loop in @jesserockz and @OttoWinter for their inputs. I'm happy to work on the TODOs & docs for this new version if we prefer it over the one in this PR.

**Comment by WeekendWarrior1** [comment]
Wow this was very fast work, great job!

> Handling arbitrary data. For this proof of concept it's just a single uint8_t, but it would be nice to support any byte array

https://stackoverflow.com/a/38914831
This thread discusses max data sizes, max payload for a BLE 4.0 write is 20 bytes, BLE 4.2 is 242 bytes.
So could you just make it a uint32_t to handle BLE 4.0 for now?
And since your value field is template-able, I would ignore handling byte arrays and let end users use a lambda if they want to do something interesting/unusual.

> Would you be interested in helping me test this out?

I don't currently have any BLE devices that I can write to so it may be a little more work to set up an environment.
I also am not every experienced with BLE lol

**Comment by jesserockz** [comment]
I think having it as just an action might be better and more useful in the long run. It allows people to set up different write values for the same characteristics etc.

**Comment by rbaron** [comment]
Thanks everyone. I'm closing this PR in favor of #3398.

**Pull request closed without merging**